### PR TITLE
Sbhuller/flx cc

### PR DESCRIPTION
**Pull request by ShyamB97**
- adds some error messages when certain elinks are not aligned
 - adds firmware tpg option to the card controller to configure the tp links

### CODE REVIEW

**Comment by wesketchum** [comment]
Tested and seems ok in recent runs at least.

File: plugins/FelixCardController.cpp [hunk 1]
```diff
@@ -81,16 +81,24 @@ FelixCardController::get_info(opmonlib::InfoCollector& ci, int /*level*/)
...
-	info.link_id = li.link_id;
-	info.enabled = li.enabled;
-	info.aligned = aligned & (1<<li.link_id);
+        std::stringstream info_name;
+        info_name << "device_" << id << "_link_" << li.link_id; 
+
+        // here we want to print out a log message when the links do not appear to be aligned.
+        // for WIB readout link_id 5 is always reserved for tp links, so alignemnt is not expected fort these
+        bool is_aligned = aligned & (1<<li.link_id);
+        if(li.link_id != 5 && !lu.emu_fanout && !is_aligned){
```
Comment (roland-sipos) [inline plugins/FelixCardController.cpp]: Please make this configurable! No hardcoded link IDs from now on please. 
I would propose to introduce a "aligned_link_mask" or "ignore_alignment_links" map through configuration.

File: src/CardControllerWrapper.cpp [hunk 2]
```diff
@@ -72,6 +72,9 @@ CardControllerWrapper::init() {
  if(bad_channels) {
     TLOG()<< bad_channels << " not aligned.";
  }
+ if(bad_channels > 2){ // Here I am assuming that the total number of links is 12 per card, so 2 are reserved for tp links and are expected to not be aligned.
```
Comment (roland-sipos) [inline src/CardControllerWrapper.cpp]: See previous comment. Bad channels: expected alignment mask - ignored alignment mask

**Review by roland-sipos — CHANGES_REQUESTED** [review]
LGTM, with elimination of hardcoded link ID.

**Comment by roland-sipos** [comment]
Also, just realized, that this change in its current form will break other front-end readouts (e.g.: DAPHNE) that utilizes all 6 channels from an SLR. So these changes are not firmware agnostic.

**Comment by ShyamB97** [comment]
so the changes I made were to add an ignore_alignment mask which is used when checking the alignment at init and get_info. The list is set in configuration and is done per SLR.
Need to test now.

**Comment by ShyamB97** [comment]
@roland-sipos @wesketchum tests on np04-srv-026 done, all looks good

File: src/CardControllerWrapper.cpp [hunk 3]
```diff
@@ -71,6 +71,11 @@ CardControllerWrapper::init() {
  int bad_channels = m_flx_card->gbt_setup( FLX_GBT_ALIGNMENT_ONE, FLX_GBT_TMODE_FEC ); //What does this do?
  if(bad_channels) {
     TLOG()<< bad_channels << " not aligned.";
+    if(bad_channels > misalignment_size){
```
Comment (roland-sipos) [inline src/CardControllerWrapper.cpp]: You check misalignment on init?

File: plugins/FelixCardController.cpp [hunk 4]
```diff
@@ -63,12 +63,17 @@ void
...
   m_cfg = args.get<felixcardcontroller::Conf>();
+  // get the number of misaligned links, this is needed since we only run init on device 0 but it checks alignment for both
+  int misalignment_size = 0;
+  for(auto lu : m_cfg.logical_units) {
+    misalignment_size += lu.ignore_alignment_mask.size();
+  }
   for (auto lu : m_cfg.logical_units) {
      uint32_t id = m_cfg.card_id+lu.log_unit_id;
      m_card_wrappers.emplace(std::make_pair(id,std::make_unique<CardControllerWrapper>(id)));
      if(m_card_wrappers.size() == 1) {
```
Comment (roland-sipos) [inline plugins/FelixCardController.cpp:74]: Misalignment will be checked only on SLR1?

**Comment by ShyamB97** [comment]
@roland-sipos  wrote separate method for link alignment check used in both init and get_info of card controller

File: scripts/flx.json [hunk 5]
```diff
@@ -1,7 +1,8 @@
 {
     "flxcardcontroller": {
         "hardware_map_file": "HardwareMap_VST.txt",
-        "emulator_mode": true
+        "emulator_mode": true,
+        "ignore_alignment_mask": [[0, 1, 2, 3, 4, 5], [0, 1, 2, 3, 4, 5]]
```
Comment (roland-sipos) [inline scripts/flx.json:5]: Shouldn't the alignment_mask be this?? Instead of the ignore one (which is empty by default)?

**Review by roland-sipos — APPROVED** [review]
LGTM, thanks

**Pull request merged**